### ISSUE TITLE
[Feature Request] Add context for Proxy

### ISSUE THREAD

**Issue by zu1k** [omjadas/hudsucker#4]
Sometimes we need to determine the `request` and then modify the `response`. 

In this case, there is a need to have a `context` to synchronize some data that is obtained from the `request` and then use them when modifying `response`.

I think we can add a `context` struct to struct `Proxy`, and pass its borrow to both `requestHandler` and `responseHandler` closures along with the `req` or `res`.

**Comment by omjadas** [comment]
I'm thinking that perhaps the request and response handlers could be combined as methods of a single struct that is passed to start_proxy. The struct would be cloned for each request/response pair. This would allow arbitrary values to be stored in the struct and used when handling the response for a specific request. Something similar to the following.

```rust
#[async_trait]
pub trait RequestResponseHandler: Send + Sync + Clone + 'static {
    async fn handle_request(&mut self, request: Request<Body>) -> RequestOrResponse;
    async fn handle_response(&mut self, request: Response<Body>) -> Response<Body>;
}

#[derive(Clone)]
struct Handler {
    pub uri: Option<Uri>,
}

#[async_trait]
impl RequestResponseHandler for Handler {
    async fn handle_request(&mut self, req: Request<Body>) -> RequestOrResponse {
        self.uri = Some(req.uri().clone());
        RequestOrResponse::Request(req)
    }

    async fn handle_response(&mut self, res: Response<Body>) -> Response<Body> {
        let uri = self.uri.unwrap();
        // Do something based on uri
        res
    }
}
```

**Comment by zu1k** [comment]
I like the new design so much, it provides a higher degree of freedom.

**Comment by omjadas** [comment]
Addressed as part of [v0.4.0](https://github.com/omjadas/hudsucker/releases/tag/v0.4.0).
